### ISSUE TITLE
Drop consult-project-extra command in favor of consult-project command?

### ISSUE THREAD

**Issue by minad** [Qkessler/consult-project-extra#2]
I open this as a separate issue since it goes beyond our last discussion. Instead of providing consult-project-extra what about just reusing consult-project-buffer and adding your sources to consult-project-buffer-sources? Also consult-project-extra-other-window is probably unnecessary given that one can use Embark. I only added consult-buffer-other-window and consult-buffer-other-frame since I assume that these commands are used often. But if I would follow a strict opinionated style, one could get rid of those commands in favor of Embark. Of course this is just a suggestion, you could as well keep your own command.

**Comment by Qkessler** [comment]
This is totally doable. I'm still thinking the pieces of this 
project, i.e, what is the plan I want to improve it with? I'm 
thinking about whether going down the road of improving the 
default project commands or focusing first on workspaces per 
project, which might also be interesting. There might be different 
ways of doing so and talking to the other package maintainers you 
mentioned in #1, I believe is the way to go.

Also moving to a new city, so life is pretty hectic now.
Take care,

On 2022-02-16, 03:29 -0800, Daniel Mendler 
***@***.***> wrote:

> I open this as a separate issue since it goes beyond our last 
> discussion.
> Instead of providing consult-project-extra what about just 
> reusing
> consult-project-buffer and adding your sources to
> consult-project-buffer-sources? Also 
> consult-project-extra-other-window is
> probably unnecessary given that one can use Embark. I only added
> consult-buffer-other-window and consult-buffer-other-frame since 
> I assume that
> these commands are used often. But if I would follow my a strict 
> opinionated
> style, one could get rid of those commands in favor of Embark.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you are subscribed to this 
> thread.


-- 
Enrique Kessler Martínez

**Comment by minad** [comment]
Sure! It makes sense to familiarize yourself completely with the existing `project-*` commands (`C-x p` prefix map) and also `consult-project-buffer`. For me this functionality is mostly sufficient, but I am sure there are features people may like, e.g., the project file source or the project root sources you have here. Have a good day and take care!

**Comment by Qkessler** [comment]
I added a quick fix on a "bug" I found. I also renamed the endpoint function to consult-project. 

Regarding the other-window command, I also use embark. Nevertheless, sometimes I like having a simple command to open the file in the other window. I use consult-buffer-other-window frequently, too. For now, I don't think the function is to be removed, I'll keep it for now and will think it through in a while. 

Finally, I was thinking about caching, as I too believe using project-files in a big repo is really slow, and therefore the consult-project function will be too. I have been looking at how projectile does caching, and I really like the approach. Have you had the change to take a look at that? Is there any talk about caching project.el in master?

This of course opens the discussion about caching inside consult-project-extra, since we are supporting emacs up to 27.1.

**Comment by minad** [comment]
> I also renamed the endpoint function to consult-project.

This is not correct. `consult-project` is outside the `consult-project-extra` namespace, which relates to our discussion in #1. What I meant in this issue is to use `corfu-project-buffer` in `corfu-project-extra`.

> For now, I don't think the function is to be removed, I'll keep it for now and will think it through in a while.

That's fine.

> Finally, I was thinking about caching, as I too believe using project-files in a big repo is really slow, and therefore the consult-project function will be too. I have been looking at how projectile does caching, and I really like the approach. Have you had the change to take a look at that? Is there any talk about caching project.el in master?

No, I am not aware of such a discussion. You may want to talk to the project.el maintainer Dmitry Gutov (cc @dgutov).

> This of course opens the discussion about caching inside consult-project-extra, since we are supporting emacs up to 27.1.

No, this is not a problem since you can install project.el from ELPA. So if you depend on a newer project.el (assuming caching has been added there) it will be pulled in by your package.

**Comment by dgutov** [comment]
Hey folks,

First of all, you might want to check whether you have a *real* big project. E.g. the two actual big projects I tested this with (Linux Kernel and Mozilla Firefox), with 70K and 200K files respectively, have their file listings within the bounds of reasonable on my machine (0.3s and 0.8s). There two examples are definitely not instant, but the projects I regularly work with are usually smaller.

If your projects are smaller too, there might be other reasons why the file listing is slow for you. E.g. you are not using the fast built-in backend (`project-vc`).

Second, caching is a fine tool, and it's listed at the top of TODO inside project.el's Commentary header. Problem is, I kind of hate the wrinkle Projectile has historically come with: the manual cache invalidation. I never see the need for it in other contemporary editors, so it feels dirty to me (non-obvious, user-unfriendly, take your pick). So the current plan is to go as far as we can without caching, and hopefully in parallel get reliable *automatic* cache invalidation (e.g. using inotify, see the same TODO entry). There had been a recursive file notification feature proposed for Emacs recently, I think it's still going through revisions/reviews.

Another approach is for people for write specialized backends for their own "big project" needs. With manual cache invalidation, if they so prefer. At least one way to do that is simply install Projectile and integrate project.el with with (the necessary 10 lines are languishing somewhere on its issue tracker).

Yet another approach would be for us to add a sort of "middleware" for file listing, which people could use to plug in their own caching. On project.el's side it would be as simple as adding a hook variable and using it in some strategic place, but all consumers would need to switch to calling some new wrapper function instead of the `project-files` generic function directly.

**Comment by minad** [comment]
I agree with your approach of avoiding caching as much as possible. However the issue of caching has come up repeatedly. For example in Consult I recently added a default configuration which determines the project root via project.el. But this leads to issues on Tramp, see https://github.com/minad/consult/commit/94224d0f91c6eddb073c812805ea3f7eaf7ed458. So maybe the root caching alone could be improved?

The file caching is orthogonal to that, I can live without file caching personally since I rely more on the recent files for quick switching. So the recent files are my second level cache, first level cache are the open buffers.

**Comment by dgutov** [comment]
Could you describe the Tramp issues? It does slow down the initial project-finding (no way around that, I think), but `project-try-vc` caches the association between the current dir and the project root.

It's probably not optimal from the cache invalidation perspective (it never happens, hence the FIXME's), but multiple calls to `project-current` from the same buffer should be cached.

**Comment by minad** [comment]
I haven't observed this personally. @astoff did (and mentioned it in https://github.com/minad/consult/commit/94224d0f91c6eddb073c812805ea3f7eaf7ed458), maybe he can give you more information.

**Comment by astoff** [comment]
> Could you describe the Tramp issues? It does slow down the initial project-finding (no way around that, I think), but project-try-vc caches the association between the current dir and the project root.

This is not what I observe, when calling `(project-root)` in a remote buffer:
- The first call takes about 2 seconds (this just tells you how bad or not my connection is to the remote host).
- A second call to `(project-current)` immediately thereafter takes about 50 ms.
- If I wait a couple of seconds, a third call to `(project-current)` again takes about 2 seconds
- If I call `(progn (project-current) (tramp-cleanup-all-connections) (benchmark-progn (project-current)))`, then a new Tramp connection is created, with a reported time of about 6 seconds for the second `(project-current)`.

From point 4 I deduce that the speedup of point 2 was due to the freshness of some network connection rather than any kind of caching.

I hope these observations are meaningful, but I must say I'm not closely acquainted with the vc, Tramp and project internals.

**Comment by dgutov** [comment]
@astoff What's your value of `project-find-functions`?

**Comment by astoff** [comment]
The default, `(project-try-vc)`. I'm using project.el version 0.8.1 on Emacs 28.

**Comment by dgutov** [comment]
Ookay, but where does `consult--project-root-default-function` come from, then?

**Comment by astoff** [comment]
I don't understand your last question, but for all effects and purposes I'm using the default `consult--project-root-default-function`.

My remarks above have nothing to do with Consult, I'm just calling `M-: (project-current)` etc on a remote buffer and observing the time it takes.

**Comment by dgutov** [comment]
Sorry, I was referring to the profile you posted in the discussion Daniel linked above. Which I have noticed only later.

I guess the behavior you are observing has to do with Tramp's caching and our calling of `vc-responsible-backend` the first thing. This should help:

```diff
diff --git a/lisp/progmodes/project.el b/lisp/progmodes/project.el
index f606a25575..e736ff41a7 100644
--- a/lisp/progmodes/project.el
+++ b/lisp/progmodes/project.el
@@ -418,29 +418,28 @@ project-vc-external-roots-function
 backend implementation of `project-external-roots'.")
 
 (defun project-try-vc (dir)
-  (let* ((backend
-          ;; FIXME: This is slow. Cache it.
-          (ignore-errors (vc-responsible-backend dir)))
+  (let* (backend
          (root
-          (pcase backend
-            ('Git
-             ;; Don't stop at submodule boundary.
-             ;; FIXME: Cache for a shorter time.
-             (or (vc-file-getprop dir 'project-git-root)
-                 (let ((root (vc-call-backend backend 'root dir)))
-                   (vc-file-setprop
-                    dir 'project-git-root
-                    (if (and
-                         ;; FIXME: Invalidate the cache when the value
-                         ;; of this variable changes.
-                         (project--vc-merge-submodules-p root)
-                         (project--submodule-p root))
-                        (let* ((parent (file-name-directory
-                                        (directory-file-name root))))
-                          (vc-call-backend backend 'root parent))
-                      root)))))
-            ('nil nil)
-            (_ (ignore-errors (vc-call-backend backend 'root dir))))))
+          (or
+           ;; FIXME: Cache for a shorter time.
+           (vc-file-getprop dir 'project-vc-root)
+           (vc-file-setprop
+            dir 'project-vc-root
+            (pcase (setq backend (ignore-errors (vc-responsible-backend dir)))
+              ('Git
+               ;; Don't stop at submodule boundary.
+               (let ((root (vc-call-backend backend 'root dir)))
+                 (if (and
+                      ;; FIXME: Invalidate the cache when the value
+                      ;; of this variable changes.
+                      (project--vc-merge-submodules-p root)
+                      (project--submodule-p root))
+                     (let* ((parent (file-name-directory
+                                     (directory-file-name root))))
+                       (vc-call-backend backend 'root parent))
+                   root)))
+              ('nil nil)
+              (_ (ignore-errors (vc-call-backend backend 'root dir))))))))
     (and root (cons 'vc root))))
 
 (defun project--submodule-p (root)
```

**Comment by dgutov** [comment]
We could also go a little further and use a buffer-local variable for this cache instead of `vc-file-getprop`. But then the cache will become buffer-local (instead of directory-local like now). Would that work for you? It would mean calling `project-current` from a new buffer will always have to recompute it.

But then invalidating it becomes quite natural: `M-x revert-buffer`.

**Comment by dgutov** [comment]
@astoff Do you have any interest in testing the patches and participating in the design discussion?

I've just pushed an improved version of above to master, but there is another change I'm mulling over.

**Comment by astoff** [comment]
@dgutov Sure, I'll test. Where is the discussion taking place?

**Comment by dgutov** [comment]
Here. :muscle: 

I've tested out the follow-up patch, and it only makes a difference when `enable-remote-dir-locals` is non-nil. I'm not sure how common that is. Do you want to try it?

Either way, it would nice if you can try and compare the performance before https://github.com/emacs-mirror/emacs/commit/86969f9658e278ebacb3d625d0309046ff1f2b54 and after.

Restart is required.

**Comment by astoff** [comment]
Hi Dmitry, I did some testing and this seems to work nicely. So what the code does is to cache the project object of each remote directory forever, right? In principle caching things forever doesn't sound great, but it's hard to imagine what could go wrong here.

I also noticed that finding a file calls `project-try-vc`. So this means that the current change will also make it faster to open the second file in any given remote directory, which is great. Now, why is `project-try-vc` being called just to find a file?

**Comment by dgutov** [comment]
> In principle caching things forever doesn't sound great, but it's hard to imagine what could go wrong here.

Things that can go wrong: someone initializes a new git repository inside current directory and sees that it's not recognized as a project root. Or deletes a .git directory and fails to see changes. The invalidation strategy could use more work, yes.

> I also noticed that finding a file calls project-try-vc.

`M-x find-file` or `M-x project-find-file`?

You might also have a mode-line package installed which uses `project-current`, but that's out of my hands.

**Comment by astoff** [comment]
> The invalidation strategy could use more work, yes.

Okay, I'm not sure a time based strategy makes a lot of sense in this case. A call to `vc-dir` is an opportunity to check if the cache is up to date. But not everyone uses vc for their version controlling.

>You might also have a mode-line package installed which uses project-current, but that's out of my hands.

Oops, of course. It doesn't happen on emacs -Q (with `M-x find-find`).

**Comment by dgutov** [comment]
> A call to vc-dir is an opportunity to check if the cache is up to date. But not everyone uses vc for their version controlling.

I was thinking of using `after-revert-hook`. And inotify might be another, more advanced option.
